open_spiel/games/checkers.cc

```diff
@@ -0,0 +1,546 @@
+        multiple_jump_piece_ = end_row * rows_ + end_column;
+      }
+      break;
+  }
+
+  if (!multiple_jump) {
+    current_player_ = 1 - current_player_; 
+  }
+
+  if (LegalActions().empty()) {
+    outcome_ = 1 - current_player_;
+  }  
+}
+
+std::string CheckersState::ActionToString(Player player,
+                                         Action action_id) const {
+  std::vector<int> values =
+      UnrankActionMixedBase(action_id, {rows_, columns_, kNumDirections, kNumMoveType, kNumPieceType, kNumPieceType});
+
+  const int start_row = values[0];
+  const int start_column = values[1];
+  const int direction = values[2];
+  const int move_type = values[3];
+  const int end_row = start_row + kDirRowOffsets[direction] * (move_type + 1);
+  const int end_column = start_column + kDirColumnOffsets[direction] * (move_type + 1);
+
+  std::string action_string =
+      absl::StrCat(ColumnLabel(start_column), RowLabel(rows_, start_row),
+                   ColumnLabel(end_column), RowLabel(rows_, end_row));
+
+  return action_string;
+}
+
+std::vector<Action> CheckersState::LegalActions() const {
+  if (moves_without_capture_ >= kMaxMovesWithoutCapture) {
+    return {};
+  }
+  std::vector<Action> move_list, capture_move_list;
+  CellState current_player_state = PlayerToState(current_player_);
+  CellState current_player_crowned = CrownState(current_player_state);
+  std::vector<int> action_bases = {rows_, columns_, kNumDirections, kNumMoveType, kNumPieceType, kNumPieceType};
+  std::vector<int> action_values = {0, 0, 0, 0, 0, 0};
+
+  for (int row = 0; row < rows_; row++) {
+    for (int column = 0; column < columns_; column++) {
+      if (BoardAt(row, column) == current_player_state || BoardAt(row, column) == current_player_crowned) {
+        for (int direction = 0; direction < kNumDirections; direction++) {
+          // Only crowned pieces can move in all 4 directions.
+          if (BoardAt(row, column) == current_player_state && ((current_player_ == 0 && direction > 1) || (current_player_ == 1 && direction < 2))) {
+            continue;
+          }
+          int adjacent_row = row + kDirRowOffsets[direction];
+          int adjacent_column = column + kDirColumnOffsets[direction];
+
+          if (InBounds(adjacent_row, adjacent_column)) {
+            CellState adjacent_state = BoardAt(adjacent_row, adjacent_column);
+            CellState opponent_state = OpponentState(current_player_state);
+            CellState opponent_state_crowned = CrownState(opponent_state);
+
+            if (adjacent_state == CellState::kEmpty) {
+              action_values[0] = row;                                 // Initial row value of player piece
+              action_values[1] = column;                              // Initial column value of player piece
+              action_values[2] = direction;                           // Direction of move for player piece
+              action_values[3] = MoveType::kNormal;                   // Type of move
```

Comment:
To make use of in UndoAction. 

Captured piece: Needed to check if player captured a normal or crowned opponent piece
Player piece: Needed because an Undo move from the last row doesn't always mean that the piece should be uncrowned